Fix xcframework embedded `.swiftmodule` file

luispadron opened the pull request:
## Summary

Some pre-built xcframeworks include several `.swiftmodule` binary files within the `.swiftmodule` directory. This happens typically when archiving with `xcodebuild`, these files are normally removed when using `xcodebuild -create-xcframework` command. However, some third party libraries include these files (not sure why) and the inclusion of these files causes the `vfs_overlay` rule to exit early and not include the `.swiftinterface` files as expected. This causes issues when attempting to import types from the given framework in Swift.

![Image from iOS](https://user-images.githubusercontent.com/13840545/180783485-185d720a-14e1-4fb4-aa40-0aa36b0d3340.jpg)

The fix here is to simply search for the top-level `.swiftmodule` **directory**. This fixes the issue within `vfs_overlay` and should continue to work for existing frameworks.

### Testing

```sh
bazel build tests/ios/xcframeworks --apple_platform_type=ios
```

The build will pass without issues.

However, before this change, if you then enable virtual frameworks:

```sh
bazel build tests/ios/xcframeworks --apple_platform_type=ios --features apple.virtualize_frameworks      
```

With this PR the above commands should pass as expected.

luispadron commented:
For some reason CI fails when adding an XCFramework test case, this possibly has something to do with Xcode 12 being used but im not super sure. I can just remove the XCFramework test case and keep the code changes, thoughts @jerrymarino?

jerrymarino commented on rules/framework/vfs_overlay.bzl:
```diff
@@ -214,7 +214,7 @@ def _make_root(vfs_parent, bin_dir_path, build_file_path, target_triple, framewo
 
 def _provided_vfs_swift_module_contents(swiftmodules, vfs_prefix, target_triple):
-    swiftmodule_file = _find_top_swiftmodule_file(swiftmodules)
+    swiftmodule_file = _find_swiftmodule_directory(swiftmodules)
```
Does this work end to end and also inside of Xcode for you after cleaning derived data with the swiftmodule?

Also if `glob` by default has [`exclude_directories`](src/main/java/com/google/devtools/build/lib/skyframe/GlobDescriptor.java) - I wonder if this is picking directories up from elsewhere ?

luispadron replied:
So it looks like the `swiftmodules` var here is passed in [framework.bzl (here)](https://github.com/bazel-ios/rules_ios/blob/f46c1406a4337c5acf3a8ea76dc89d4eee5271df/rules/framework.bzl#L219-L229). These `framework_files` are gathered [here](https://github.com/bazel-ios/rules_ios/blob/master/rules/framework.bzl#L246)

luispadron replied:
See this comment: https://github.com/bazel-ios/rules_ios/pull/521#discussion_r929851643

jerrymarino reviewed (commented):
Thanks for the PR @luispadron !

jerrymarino commented:
@luispadron - re: Xcode 13 - based on what the CI is emitting it looks like the new test case is failing because of the concurrency framework. If so - maybe just remove the dep to detangle the Xcode 13 upgrade from it: 
> /private/var/tmp/_bazel_runner/c2342bca25695567c8c3f3be166b56b3/execroot/build_bazel_rules_ios/tests/ios/xcframeworks/BasicXCFramework/BasicXCFramework.xcframework/ios-arm64_x86_64-simulator/BasicXCFramework.framework/Modules/BasicXCFramework.swiftmodule/x86_64-apple-ios-simulator.swiftinterface:7:8: error: no such module '_Concurrency'
import _Concurrency
       ^
tests/ios/xcframeworks/Sources/Foo.swift:2:8: error: failed to build module 'BasicXCFramework' from its module interface; the compiler that produced it, 'Apple Swift version 5.5.2 (swiftlang-1300.0.47.5 clang-1300.0.29.30)', may have used features that aren't supported by this compiler, 'Apple Swift version 5.4.2 (swiftlang-1205.0.28.2 clang-1205.0.19.57)'
import BasicXCFramework
       ^
swift_worker: Could not copy bazel-out/ios-x86_64-applebin_ios-ios_x86_64-dbg-ST-2ee1e31a3506/bin/_swift_incremental/tests/ios/xcframeworks/xcframeworks_swift_objs/Sources/Foo.swift.o to bazel-out/ios-x86_64-applebin_ios-ios_x86_64-dbg-ST-2ee1e31a3506/bin/tests/ios/xcframeworks/xcframeworks_swift_objs/Sources/Foo.swift.o (No such file or directory)

jerrymarino commented:
The other thing about this PR as it sits right now is it requires the minimum Xcode version you used to generate it because of this .swiftinterface. I hoped ABI stability would help this: did you try to turn on stable ABI mode in Xcode?

Also, I assumed you generated this fixture Xcode, would you kindly check in the project and add a README of the steps you took in the directory to generate the `.xcframework` ? That will make it considerably less work for devs to iterate on it a bit.

jerrymarino commented:
On one env I get this:
```
d2b3b8d4835/bin/tests/ios/xcframeworks/xcframeworks.swiftmodule-0.params)
# Configuration: 1f904adaabb40ab7c5ec17f85da239fb698a2c452c71b9febe15ee3d2557e84b
# Execution platform: @local_config_platform//:host
/private/var/tmp/_bazel_jmarino/4180ce3012e050416efacbaebca2f601/execroot/build_bazel_rules_ios/tests/ios/xcframeworks/Sources/Foo.swift:2:8: error: module 'BasicXCFramework' was created for incompatible target x86_64-apple-ios15.2-simulator: /private/var/tmp/_bazel_jmarino/4180ce3012e050416efacbaebca2f601/execroot/build_bazel_rules_ios/tests/ios/xcframeworks/BasicXCFramework/BasicXCFramework.xcframework/ios-arm64_x86_64-simulator/BasicXCFramework.framework/Modules/BasicXCFramework.swiftmodule/x86_64.swiftmodule
import BasicXCFramework
       ^
swift_worker: Could not copy bazel-out/macos-x86_64-applebin_macos-darwin_x86_64-dbg-ST-cd2b3b8d4835/bin/_swift_incremental/tests/ios/xcframeworks/xcframeworks_swift_objs/Sources/Foo.swift.o to bazel-out/macos-x86_64-applebin_macos-darwin_x86_64-dbg-ST-cd2b3b8d4835/bin/tests/ios/xcframeworks/xcframeworks_swift_objs/Sources/Foo.swift.o (No such file or directory)
Target //tests/ios/xcframeworks:xcframeworks failed to build
INFO: Elapsed time: 3.158s, Critical Path: 3.01s
INFO: 2 processes: 2 internal.
FAILED: Build did NOT complete successfully

```

luispadron commented:
@jerrymarino yeah Im going to try and mess with the build setting and recreate the xcframework to see if we can get this passing with Xcode 12. 

As far as, how I created it. I used the typical `xcodebuild archive` command to create two frameworks (one for ios SDK and one for simulator SDK). I then used `xcodebuild -create-xcframework` to create the xcframework which includes both the above frameworks. 

However, in order to replicate the .swiftmodule issue i then copied the `.swiftmodule` files from the frameworks archives into the individual xcframework `.swiftmodule` directories. 

This mimicked the format of the Plaid SDK I was testing without having to include that large SDK here for testing purposes. I'm not sure why that SDK is copies those files back in or how it's being built to have this format but I wanted to add a fixture here that surfaced that issue. 

In terms of adding it to the readme, where should that go, and what exactly are you looking to add there?

luispadron commented:
Found this with some more info on the implicit dependency on currency when using Xcode 13 to build: https://developer.apple.com/forums/thread/690350

I've updated the xcframework fixture with that setting and it looks to have removed `import _Concurrency` from the `.swiftinterface` files so think this should allows us to build with Xcode 12 in CI.

luispadron commented on rules/framework/vfs_overlay.bzl:
```diff
@@ -37,9 +37,9 @@ def _get_vfs_parent(ctx):
+def _find_swiftmodule_directory(swiftmodules):
     for file in swiftmodules:
-        if file.basename.endswith(".swiftmodule"):
+        if file.basename.endswith(".swiftmodule") and file.is_directory:
```
I'm really not familiar with all the edge cases or how the format is supposed to work here. 

From what I understood of the original intent here this was to find the "top-level" swiftmodule file (previously this was either a directory or a source file). The code that uses this then does different things for source vs. directory which this change breaks I believe. 

Perhaps what we can do is:
- Attempt to find any `.swiftmodule` file
- Once found, continue searching
- If we find a `.swiftmodule` directory then we return that
- Otherwise we return the original file we found in step #1

This should keep the original intent while preferring `.swiftmodule` directories which I believe should fix the issue here.

luispadron replied:
```diff
-            swiftmodule_file = _find_top_swiftmodule_file(swiftmodules)
-            if not swiftmodule_file and parent_dir_base_name.endswith(".swiftmodule"):
+            if parent_dir_base_name.endswith(".swiftmodule"):
```

I believe this change makes more sense and I verified it fixes the issue here with the VFS overlay. I've also tested it without the VFS feature in both terminal and via the `xcodeproj` rule. Let me know your thoughts!

jerrymarino replied:
Ok glad you caught it, the new way seems to make sense so far. This code is very caveated and needs to work end to end with programs which are reading the outputs. I'm going to verify 1-2 edge cases that were left unchecked 👍

jerrymarino reviewed (commented):
Good to hear the lower version produced a more compatible xcframework!

 The only suggestion I've got is to check in this xcodeproj project. Creating an xcframework end to end with Xcode is unknown to me and also relies on closed source Xcode stuff: if we need to tweak your first example a bit as a follow and I'm happy to help. 

In general we may want stable ABI on this example, but we can work towards this as a followup.


> I've updated the xcframework fixture with that setting and it looks to have removed import _Concurrency from the .swiftinterface files so think this should allows us to build with Xcode 12 in CI.

This new toolchain is just riddled with dependencies on concurrency / breaking old devices 💰

luispadron commented:
> The only suggestion I've got is to check in this xcodeproj project. Creating an xcframework end to end with Xcode is unknown to me and also relies on closed source Xcode stuff: if we need to tweak your first example a bit as a follow and I'm happy to help.

@jerrymarino I've added a commit adding the Xcode project I used to create the XCFramework. I also added a README file with the command used to generate the xcframework
